spock-core/src/main/java/org/spockframework/runtime/extension/builtin/TagExtension.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.spockframework.runtime.extension.builtin;
+
+import org.spockframework.runtime.extension.IAnnotationDrivenExtension;
+import org.spockframework.runtime.model.FeatureInfo;
+import org.spockframework.runtime.model.SpecElementInfo;
+import org.spockframework.runtime.model.SpecInfo;
+import spock.lang.Tag;
+
+/**
+ * Processes {@link Tag @Tag} annotations on specifications and features, converting them to
+ * {@link org.spockframework.runtime.model.Tag Spock tags} which can potentially be used by reporting extensions.
+ * More importantly, it also makes the tag information available to JUnit 5 as
+ * {@link org.junit.platform.engine.TestTag platform tags} in order to enable filtering, e.g. from build tools like
+ * Maven or Gradle.
+ *
+ * @see <a href="https://junit.org/junit5/docs/current/user-guide/#running-tests-tag-expressions">
+ *   JUnit platform tag expression syntax</a>
+ * @author Alexander Kriegisch
+ * @since 2.2
+ */
+public class TagExtension implements IAnnotationDrivenExtension<Tag> {
+  public static final String TAG_EXTENSION_KEY = "TagExtension";
+
+  @Override
+  public void visitSpecAnnotation(Tag tag, SpecInfo spec) {
+    addTags(tag, spec);
```

Comment:
Can you elaborate, please? I am not sure I understand.

Reply:
Annotation driven extensions are only executed for the Spec they are defined upon, since Spock models inheritance by having a double linked list of `SpecInfo` an annotation defined on a parent Spec will not modify the `SpecInfo` that is used by the `SpecNode` in the end, so if you want to inherit then you need to use `spec.getBottomSpec()` to put the tags into the `SpecInfo` instance that is actually used by `SpecNode`.

Reply:
Oh, now I understand what you mean. The problem with using the bottom spec is that in the case of
* `A extends Specification` (`A` being abstract or maybe not, it does not matter) and
* `B extends A`,

feature methods inherited from `A` will **not** be executed in `B`, if the tags from `A` are only added to `B` (bottom spec), as you suggest.

I really think we should stick to the way I implemented it, i.e. adding tags to the visited items (here: specs), not to something further up or down the inheritance chain, and then later walking that chain. The extension, like I said before, is agnostic of how tags are going to be used later. It simply makes sure to convert each annotation to a corresponding tag. Whoever consumes those tags can then decide if/how to aggregate them. `SpockNode` knows this, because it responds to tag requests from the JUnit platform.

Reply:
Please note that 886dad34 moves aggregation to `ITaggable` and its implementing classes. The commit comment explains in detail why I did that and what the benefits are. For `TagExtension` it means that we do not need to aggregate anything here, avoiding to scatter duplicate tags across the spec hierarchy and each single feature. I am resolving this conversation. Please reopen if you disagree.